lib/elixir/lib/module/types/descr.ex

```diff
@@ -188,6 +244,49 @@ defmodule Module.Types.Descr do
 
   defp iterator_difference(:none, map), do: map
 
+  ## Type relations
+
+  @doc """
+  Check if a type is a subtype of another.
+
+  If     `left  ~= (left_dyn  and dynamic()) or left_static`
+  and    `right ~= (right_dyn and dynamic()) or right_static`
+
+  then the gradual subtyping relation defined in Definition 6.5 page 125 of
+  the thesis https://vlanvin.fr/papers/thesis.pdf is:
+
+  `left <= right` if and only if
+    - `left_static <= right_static`
+    - `left_dyn <= right_dyn`
+
+  Because of the dynamic/static invariant in the `descr`, subtyping can be
+  simplified in several cases according to which type is gradual or not.
+  """
+  def subtype?(%{} = left, %{} = right) do
+    is_grad_left = gradual?(left)
+    is_grad_right = gradual?(right)
+
+    cond do
+      is_grad_left and not is_grad_right ->
+        left_dynamic = Map.get(left, :dynamic)
+        subtype_static(left_dynamic, right)
+
+      is_grad_right and not is_grad_left ->
+        right_static = Map.delete(right, :dynamic)
+        subtype_static(left, right_static)
+
+      true ->
+        subtype_static(left, right)
+    end
+  end
+
+  defp subtype_static(left, right), do: empty?(difference_static(left, right))
+
+  @doc """
+  Check if a type is equal to another.
+  """
+  def equal?(left, right), do: subtype?(left, right) and subtype?(right, left)
```

Comment:
Why can't we compare the descrs directly (after normalizing the double representation of a static type), do we expect more cases of multiple representations?

Reply:
You're correct. Right now, this could be avoided by normalizing gradual types, but tuple types will have multiple representations. For instance, the type that describes either pairs of an integer with any value, or pairs of booleans:
`{int, term} or {bool, bool}`. 
can equivalently be represented as:
`{int or bool, bool} or {int, term and not bool}`.

I don't know which one to pick in general, and determining if a unique representation exists for all tuple types is, I believe, an open question. So, we support representing both and will use `equal?` if we need to check that they are semantically equal.

Reply:
In practice, I don't think equality will be used in the type-checker. It's always subtyping. Since it's only used for tests, I'm thinking of defining it there.

Reply:
Makes sense! Not for me to say, but I think having it in this module is fine :)